src/Blammo/Logging/Terminal.hs

```diff
@@ -21,57 +21,82 @@ module Blammo.Logging.Terminal
 import Prelude
 
 import Blammo.Logging.Colors
+import Blammo.Logging.Terminal.LogPiece (LogPiece, logPiece)
+import qualified Blammo.Logging.Terminal.LogPiece as LogPiece
 import Control.Monad.Logger.Aeson
 import Data.Aeson
 import Data.Aeson.Compat (KeyMap)
 import qualified Data.Aeson.Compat as Key
 import qualified Data.Aeson.Compat as KeyMap
 import Data.ByteString (ByteString)
 import qualified Data.ByteString.Lazy as BSL
+import Data.List (sortOn)
 import Data.Maybe (fromMaybe)
 import Data.Text (Text, pack)
 import qualified Data.Text as T
-import Data.Text.Encoding (encodeUtf8)
 import Data.Time (defaultTimeLocale, formatTime)
 import qualified Data.Vector as V
 
-reformatTerminal :: Bool -> LogLevel -> ByteString -> ByteString
-reformatTerminal useColor logLevel bytes = fromMaybe bytes $ do
+reformatTerminal :: Int -> Bool -> LogLevel -> ByteString -> ByteString
+reformatTerminal breakpoint useColor logLevel bytes = fromMaybe bytes $ do
   LoggedMessage {..} <- decode $ BSL.fromStrict bytes
 
   let
     colors@Colors {..} = getColors useColor
 
-    logTimestampText =
-      dim $ pack $ formatTime defaultTimeLocale "%F %X" loggedMessageTimestamp
+    logTimestampPiece = logPiece dim $ pack $ formatTime
+      defaultTimeLocale
+      "%F %X"
+      loggedMessageTimestamp
 
-    logLevelText = case logLevel of
-      LevelDebug -> gray $ padTo 9 "debug"
-      LevelInfo -> green $ padTo 9 "info"
-      LevelWarn -> yellow $ padTo 9 "warn"
-      LevelError -> red $ padTo 9 "error"
-      LevelOther x -> blue $ padTo 9 x
+    logLevelPiece = case logLevel of
+      LevelDebug -> logPiece gray $ padTo 9 "debug"
+      LevelInfo -> logPiece green $ padTo 9 "info"
+      LevelWarn -> logPiece yellow $ padTo 9 "warn"
+      LevelError -> logPiece red $ padTo 9 "error"
+      LevelOther x -> logPiece blue $ padTo 9 x
 
     loggedSourceAsMap =
       foldMap (KeyMap.singleton "source" . String) loggedMessageLogSource
 
-  pure $ encodeUtf8 $ mconcat
-    [ logTimestampText <> " "
-    , "[" <> logLevelText <> "] "
-    , bold $ padTo 31 loggedMessageText
-    , colorizeKeyMap colors loggedSourceAsMap
-    , colorizeKeyMap colors loggedMessageThreadContext
-    , colorizeKeyMap colors loggedMessageMeta
-    ]
-
-colorizeKeyMap :: Colors -> KeyMap Value -> Text
-colorizeKeyMap Colors {..} km
-  | KeyMap.null km = ""
-  | otherwise = " " <> T.intercalate " " keyValues
- where
-  keyValues = map (uncurry renderPair) $ KeyMap.toList km
+    logPrefixPiece =
+      mconcat [logTimestampPiece <> " ", "[" <> logLevelPiece <> "] "]
```

Comment:
Why not?
```suggestion
      logTimestampPiece <> " [" <> logLevelPiece <> "] "
```

Reply:
This was just personal preference, but I have heard that `mconcat` is often faster than repeated `(<>)` (it depends on the instance, of course).

Reply:
Oh I didn't notice that, this was a copy/paste from when I was doing,

```hs
mconcat
  [ a <> b <> c
  , d <> e <> f
  ]
```

In which case it may make more sense because I wanted `a <> b <> c` to feel like one logical piece in a list of pieces.

Reply:
Like compare,

```hs
key
  <> "="
  <> value
  <> ", "
  <> keyB
  <> "="
  <> "valueB"
```

With,

```hs
mconcat
  [ key <> "=" <> value <> ", "
  , keyB <> "=" <> valueB
  ]
```